pinot-core/src/main/java/com/linkedin/pinot/core/data/readers/sort/PinotSegmentSorter.java

```diff
@@ -0,0 +1,119 @@
+/**
+ * Copyright (C) 2014-2016 LinkedIn Corp. (pinot-core@linkedin.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.linkedin.pinot.core.data.readers.sort;
+
+import com.linkedin.pinot.common.data.FieldSpec;
+import com.linkedin.pinot.common.data.Schema;
+import com.linkedin.pinot.core.data.readers.PinotSegmentColumnReader;
+import it.unimi.dsi.fastutil.Arrays;
+import it.unimi.dsi.fastutil.Swapper;
+import it.unimi.dsi.fastutil.ints.IntComparator;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+
+/**
+ * Sorter implementation for pinot segments
+ */
+public class PinotSegmentSorter implements SegmentSorter {
+
+  private int _numDocs;
+  private Schema _schema;
+  private Map<String, PinotSegmentColumnReader> _columnReaderMap;
+  private Map<String, Map<Object, Integer>> _mvColumnDictionaries;
+  private int[] _sortOrder;
+  private List<String> _dimensionNames;
+  int _numDimensions;
+
+  public PinotSegmentSorter(int numDocs, Schema schema, Map<String, PinotSegmentColumnReader> columnReaderMap) {
+    _numDocs = numDocs;
+    _schema = schema;
+    _columnReaderMap = columnReaderMap;
+    _dimensionNames = new ArrayList<>();
+    for (FieldSpec fieldSpec : _schema.getAllFieldSpecs()) {
+      // Count all fields that are not metrics as dimensions
+      if (fieldSpec.getFieldType() != FieldSpec.FieldType.METRIC) {
+        String dimensionName = fieldSpec.getName();
+        _numDimensions++;
+        _dimensionNames.add(dimensionName);
+      }
+    }
+  }
+
+  /**
+   * Sort the segment by the sort order columns. For PinotSegmentSorter, orderings are computed by comparing
+   * dictionary ids.
+   *
+   * TODO: add the support for no-dictionary and multi-value columns.
```

Comment:
Is star-tree supported? If not, please list here. Also, all unsupported features should be caught and bailed out upfront.

Reply:
Startree index generation will simply call `recordReader.nex()` so it should work with the startree. Currently, no-dictionary and multi value columns are features that are not supported. I will catch these cases upfront at the `PinotSegmentSorter`.